sui-execution/src/executor.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    collections::{BTreeSet, HashSet},
+    sync::Arc,
+};
+
+use sui_protocol_config::ProtocolConfig;
+use sui_types::{
+    base_types::{ObjectRef, SuiAddress, TxContext},
+    committee::EpochId,
+    digests::TransactionDigest,
+    effects::TransactionEffects,
+    error::ExecutionError,
+    execution::{ExecutionState, TypeLayoutStore},
+    execution_mode::ExecutionResult,
+    gas::SuiGasStatus,
+    metrics::LimitsMetrics,
+    temporary_store::{InnerTemporaryStore, TemporaryStore},
+    transaction::{ProgrammableTransaction, TransactionKind},
+    type_resolver::LayoutResolver,
+};
+
+/// Abstracts over access to the VM across versions of the execution layer.
+pub trait Executor {
```

Comment:
Originally, I intended to expose the `ExecutionMode` trait on this API boundary, but that's not possible because the `Executor` must be passed as a trait object, so the trait can't have generic type arguments (of which the `Mode` would have been one).  Instead the modes manifest as 4 different functions:

- `execute_transaction_to_effects` (`Normal`).
- `execute_system_transaction_to_effects` (`System`) -- potentially unused.
- `dev_inspect_transaction` (`DevInspect`).
- `update_genesis_state` (`Genesis`) -- runs a single PTB, rather than a `TransactionKind`.

Reply:
Do you know why ExecutionMode was defined as a trait instead of just an Enum?

Reply:
`ExecutionMode` comes with associated types, so it wouldn't be possible to encode it as an enum.  This is because when dev-inspecting a transaction, the return type of `execute_transaction_to_effects` changes.

Reply:
got to say that I always wondered about `Mode` and whether code refactoring could get rid of it. And this feels already a partial refactor, if nothing else exposing different functions for the specific case up the call stack.
I am obviously not asking to change anything but I wonder if you have an opinion about it

Reply:
The main thing that makes this difficult is dev inspect transactions which behave significantly differently (do extra work tracking intermediate results and changes to inputs, and return a different kind of execution result).  This makes sense because that's what `ExecutionMode` was originally created for.  My take is that the complexity here is inherent, and it's not worth trying to replace`ExecutionMode`, you'll just get something else equally complex.